File: irmin-connectors/models/connector.go
package models

import (
	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// ConnectorCapability describes the features the connector supports.
type ConnectorCapability string

const (
	// Can perform operation "pull"
	ConnectorCapabilityPullFullSync ConnectorCapability = "pull"
	// Can perform operation "push"
	ConnectorCapabilityPushFullSync ConnectorCapability = "push"
	// Can send patch webhook events
	ConnectorCapabilityPushPatchSync ConnectorCapability = "webhook_patch"
	// Can send full object webhook events
	ConnectorCapabilityPullPatchSync ConnectorCapability = "webhook_pull"
)

// ConnectorCategory describes the type of service the connector is for.
type ConnectorCategory string

const (
	ConnectorCategoryDatabase          ConnectorCategory = "database"
	ConnectorCategoryCRM               ConnectorCategory = "crm"
	ConnectorCategoryERP               ConnectorCategory = "erp"
	ConnectorCategoryWarehouse         ConnectorCategory = "warehouse"
	ConnectorCategoryMarketing         ConnectorCategory = "marketing"
	ConnectorCategoryAnalytics         ConnectorCategory = "analytics"
	ConnectorCategoryStorage           ConnectorCategory = "storage"
	ConnectorCategoryMessaging         ConnectorCategory = "messaging"
	ConnectorCategoryPayment           ConnectorCategory = "payment"
	ConnectorCategorySocial            ConnectorCategory = "social"
	ConnectorCategoryCalendar          ConnectorCategory = "calendar"
	ConnectorCategoryProjectManagement ConnectorCategory = "project_management"
	ConnectorCategoryECommerce         ConnectorCategory = "ecommerce"
	ConnectorCategoryIoT               ConnectorCategory = "iot"
	ConnectorCategoryMonitoring        ConnectorCategory = "monitoring"
	ConnectorCategoryOther             ConnectorCategory = "other"
)

// ConnectorInfo holds general information about a connector.
type ConnectorInfo struct {
	gorm.Model

	// Unique identifier for the connector
	ID string `json:"id" gorm:"primaryKey type:varchar(255);not null"`

	// Name of the connector
	Name string `json:"name" gorm:"type:varchar(255);not null"`

	// Short description of the connector
	Description string `json:"description" gorm:"type:text;not null"`

	// Current version of the connector
	Version string `json:"version" gorm:"type:varchar(50);not null"`

	// Version of the Irmin Connector Structure this connector adheres to
	StructureVersion string `json:"structure_version" gorm:"type:varchar(50);not null"`

	// Name of the author of the connector
	Author string `json:"author" gorm:"type:varchar(100);not null"`

	// Base URL for the connector's REST API
	APIBaseURL string `json:"api_base_url" gorm:"type:text;not null"`

	// URL to the connector's logo image
	LogoURL string `json:"logo_url" gorm:"type:text;not null"`

	// List of capabilities supported by the connector
	// Stored in JSON format for array-like data.
	Capabilities datatypes.JSON `json:"capabilities" gorm:"type:json"`

	// List of locales supported by the connector
	// Also stored in JSON format.
	Locales datatypes.JSON `json:"locales" gorm:"type:json"`

	// (optional) Primary category of the connector
	PrimaryCategory *ConnectorCategory `json:"primary_category" gorm:"type:varchar(50)"`

	// (optional) List of categories the connector belongs to
	Categories datatypes.JSON `json:"categories" gorm:"type:json"`

	// (optional) Email address of the author
	AuthorEmail *string `json:"author_email" gorm:"type:varchar(255)"`

	// (optional) Markdown-formatted text providing more details about the connector
	Documentation *string `json:"documentation" gorm:"type:text"`

	// (optional) URL to read more about the connector, such as documentation
	ReadMoreURL *string `json:"read_more_url" gorm:"type:text"`
}

// TableName allows you to override the default table name if desired.
func (ConnectorInfo) TableName() string {
	return "connector_info"
}

// Generate a custom string ID before creating the record.
func (ci *ConnectorInfo) BeforeCreate(tx *gorm.DB) (err error) {
	if ci.ID == "" { // Only generate an ID if it wasn't already set
		id := uuid.NewString()
		ci.ID = id
	}
	return nil
}

File: irmin-ai/src/irmin-api/types/Connector.ts
/**
 * Irmin connector type
 */
export interface Connector {
  /** ID is a unique SQID identifier of the connector */
  id: string;
  /** Name of the connector */
  name: string;
  /** Short description of the connector */
  description: string;
  /** Current version of the connector (provided by the connector author) */
  version: string;
  /** Version of the Irmin Connector Structure this connector adheres to (e.g. which connector guideline version the connector is based on) */
  structure_version: string;
  /** Name of the author of the connector */
  author: string;
  /** URL to the connector's logo image */
  logo_url: string;
  /** List of capabilities supported by the connector, eg. what kind of operations the connector can perform */
  capabilities: ConnectorCapability[];
  /** List of locales supported by the connector, eg. what languages the connector supports */
  locales: string[];
  /** Array of categories associated with the connector, eg. what kind of connector it is */
  categories: ConnectorCategory[];
  /** Primary category of the connector */
  primary_category: ConnectorCategory;
  /** Email address of the author, eg. who to contact if there are any issues with the connector */
  author_email: string;
  /** URL to read more about the connector, eg. a website where the connector is described, such as documentation */
  read_more_url: string;
}

/**
 * Represents the capabilities of a connector.
 *
 * 'pull' means that data files can be pulled from different paths in the connector
 * 'push' means that data files can be pushed to different paths in the connector
 * 'apply_patch' means that the connector can receive and apply patch operations
 * 'patch_event' means that the connector can emit patch events via webhooks when data changes
 */
type ConnectorCapability = 'pull' | 'push' | 'apply_patch' | 'patch_event';

/**
 * Represents the category of a connector.
 */
type ConnectorCategory =
  | 'analytics'
  | 'calendar'
  | 'crm'
  | 'database'
  | 'ecommerce'
  | 'erp'
  | 'iot'
  | 'marketing'
  | 'messaging'
  | 'monitoring'
  | 'other'
  | 'payment'
  | 'project_management'
  | 'social'
  | 'storage'
  | 'warehouse';
